### PR TITLE
[#11] feature: model and repository of creating game

### PR DESCRIPTION
**Pull request by teds-lin**
實作以下部份並且新增相關測試案例：

- Domain models
  - Game
   -`add_players()` 
     * 玩家人數限制為2~4人
     * bind Player instances to specific Game instance
  - `id` , identifier could be generated in UUID v4 format
  - `assign_character()`, 僅紀錄目前哪些調查員已被選擇，具體那個玩家選擇那個調查員由`Player class`紀錄

- Player
  - 紀錄選擇那個調查員

- In-memory repository
  - `InMemoryRepository.save()` 
    - Using hash table to store key (game ID) value (relevant Game instance) pairs
    - Provide thread-safe access to the in-memory games dictionary

### CODE REVIEW

File: backend/app/usecase/create_game.py [hunk 1]
```diff
@@ -17,19 +17,21 @@ async def execute(self, data: List[PlayerDto]) -> CreateGameRespDto:
...
         return CreateGameRespDto(url=url)
 
     def rand_select_investigator(self, game: Game) -> List[GameError]:
-        options = random.shuffle(list(Investigator))
+        options = list(Investigator)
+        random.shuffle(options)
 
         def fn1(p):
             c = options.pop(0)
-            return game.assign_character(p, c)
```
Comment (metalalive) [inline backend/app/usecase/create_game.py:32]: Initially my thought was to assign an investigator to a player completely at **domain layer** ,
but I  think it is also OK to do that at **use-case layer**.

btw ,Ted, you don't need to change anything at here, it is just discussion.

**Review by metalalive — COMMENTED** [review]
Ted thanks for the code, the domain models look good,
I have questions about in-memory repository , and left my review comments at `files changed` tab, please take a look, thanks

c.c. @wingtkw , If you'd like to learn the code in this PR, please leave comments at here or in [GitHub Discussion](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/discussions)

File: backend/app/adapter/repository/in_mem.py [hunk 2]
```diff
@@ -1,8 +1,37 @@
...
+    def games_dict(self) -> Dict[str, Game]:
+        """Provide thread-safe access to the in-memory games dictionary.
+
+        This context manager acquires a lock before yielding the games dict
+        and releases it after, ensuring thread-safe access.
+
+        Yields:
+            Dict[str, Game]: The in-memory games dictionary.
+        """
+        lock = threading.Lock()
```
[the remark this answers is not in the document]
Comment (teds-lin) [reply backend/app/adapter/repository/in_mem.py:22]: > However the code here looks like every thread creates a new lock and acquires the lock for each call to the function.

確實，這邊我稍微修正一下

> `threading.Lock()`

寫的時候沒想清楚，其他地方都是用`asyncio`套件，我不該混用`thread`(查資料是說雖然可以，但盡量避免)，我將改寫，此處變更為` asyncio.Lock()`

File: backend/app/usecase/create_game.py (hunk 1 above)
Comment (teds-lin) [reply backend/app/usecase/create_game.py:32]: 目前只是一個簡單的分配，後續若有較多複雜業務邏輯需處理，在從use-case layer移至domain layer

File: backend/app/adapter/repository/in_mem.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/repository/in_mem.py]: Thanks for reminding, `asyncio.Lock` would be more appropriate in such case

**Review by metalalive — APPROVED** [review]
Looks good to me, thanks again

File: backend/app/adapter/repository/in_mem.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (metalalive) [reply backend/app/adapter/repository/in_mem.py]: Thanks for reminding, `asyncio.Lock()` is more appropriate in the case that multiple asynchronous tasks acquire the lock concurrently.

**Pull request merged**